### DISCUSSION TITLE
How to match a custom type/interface?

### DISCUSSION THREAD

**Discussion by KlausJokisuo** [gvergnaud/ts-pattern#115] [Q&A]
Heya, 
I was wondering how I could check if the match is the correct type.

I know that I can check if the match is null, but I would like to check the correct type too.

```ts
type User = {
  age: number;
  name: string;
  uid: string;
};

type NullOrUser = User | null;

const nullOrUser: NullOrUser = {
  age: 10,
  name: 'Fuu',
  uid: 'Faa',
};

match(nullOrUser)
  .with(P.checkTypeIsUser, () => console.log('It´s User')) // <--- This line here
  .with(P.nullish, () => console.log('It´s null'))
  .exhaustive();
```

Thanks 👍

**Comment by gvergnaud** [comment]
Hey!

The recommended way to achieve this is to create a user pattern and then infer the type from the pattern: 

```ts
const userPattern = {
    age: P.number,
    name: P.string,
    uid: P.string,
};

type User = P.infer<typeof userPattern>;

type NullOrUser = User | null;

const nullOrUser: NullOrUser = {
    age: 10,
    name: 'Fuu',
    uid: 'Faa',
};

match(nullOrUser)
    .with(userPattern, () => console.log('It´s User'))
    .with(P.nullish, () => console.log('It´s null'))
    .exhaustive();
```

**Reply by KlausJokisuo** [reply]
Thanks!

**Reply by kturmel** [reply]
How can you then P.select in your custom pattern?

Context

```ts
type None = {
  _tag: 'None'
}

type Some<T> = {
  _tag: 'Some'
  value: T
}

type Option<T> = Some<T> | None
```

Example

```ts
const optionPattern = {
  _tag: 'Some',
  value: P.any
}

const some = <T>(v: T): Option<T> => ({ _tag: 'Some', value: v })

match(some(2))
  .with(optionPattern, (v) => ...) // v is { _tag: 'Some', value: 2 } or shortly Some<2>
  .otherwise(() => ...)
```

How can I select the value from the option then ?

```ts
match(some(2))
  .with({ _tag: 'Some', value: P.select() }, (v) => ...) // v is 2
  .otherwise(() => ...)
```

Like this but with custom pattern? Do I have to repeat myself the `_tag: 'Some'` everytime?

_yes, this is some kind of topic necro_

**Reply by gvergnaud** [reply]
Your solution seems to be working if the tag key is defined as a constant:

```ts
type Option<T> = Some<T> | None

const optionPattern = {
  _tag: 'Some' as const, // 👈 added `as const`
  value: P.select()
}

const some = <T>(v: T): Option<T> => ({ _tag: 'Some', value: v })

match(some(2))
  .with(optionPattern, (v) => { v + 1 }) // v: number
  .otherwise(() =>  {})
```

[Playground](https://www.typescriptlang.org/play?jsx=0&ssl=23&ssc=24&pln=12&pc=1#code/JYWwDg9gTgLgBAbxAQxgYwBYBo4AUC+cAZlBCHAOQwDOAtGKjAKZQB2FAUBzAJ5hNwAchFYCAvIg5w4AfRjIA5gC5Kw0Z3xde-OAGUyTADwAVAHxwJCKbPnLK+kE07SAbsgA2AVyYrjHTdx8AgDyYDDAIibmEg5GZnAAPkIiTFxoItTwEGERrLiMLKwWktJyiioUsRRwyNRw6ayZWNZuXj54AHTUTO5MaDAAFACU-mkZ8NQGxVEDLr5DKqHhkfFi5gMINuX2BhQ4rd4qLnD4Ixwo6BgDk44DAExDI9IdAO7AMFfZy3kFbDizQws5k2xwA1HAAIwnQEAehhcDmcFYnhAACMWNYOhAPiw3t0BsMgdIEKcgA)

**Reply by kturmel** [reply]
Oh this is banger!

Thanks for your solution.
